### PR TITLE
Feature | Force option for either Basic or Advanced Query Builder

### PR DESCRIPTION
**Pull request by pauline-n**
Feature | force option for either Basic or Advanced Query Builder. #632

### CODE REVIEW

**Comment by wakibi** [comment]
Hi @pauline-n, please ensure you test (and by extension lint) your code well before marking a PR as ready. It will save all of us valuable time. Make sure it's passing the tests at the very least.

**Comment by pauline-n** [comment]
Hi @wakibi , sorry about that. Working on that now then will re-request review.

Thanks
Pauline

**Comment by ThatcherK** [comment]
@pauline-n Functional review

- [x] The advanced query builder is not selected by default
- [x] Am not comfortable with the labelling of the save button, could we label it along the lines of 'go to querybuilder' instead? For example when I don't check 'remember my choice', I won't know what to click next because I am not saving anything, yet the save button is the same that takes you to the query-builder. It's not very intuitive when it's labelled save.
- [x] On selection of a query builder of choice and checking the 'remember my choice' checkbox and saving, am taken to the desired querybuilder. But when I leave the querybuilder and and then try to access the querybuilder again via the sidebar menu, am given the modal pop -up again.I thought the choice is supposed to be saved. It looks like that isn't working

**Comment by pauline-n** [comment]
@ThatcherK  thank you for the feedback, I must have missed out some steps let me get them working and will push commits afterwards

**Comment by pauline-n** [comment]
Hello @ThatcherK have worked on the missing details above. Kindly review, all should be good now.

Thanks
Pauline

**Comment by ThatcherK** [comment]
@ Pauline still on the functional side, 

- [x] When I select a query builder of choice, and check remember me, while still logged in, my choice is remembered, when I move away from the query builder , then later go back, am taken to the query builder I selected. Which is working perfectly well. The problem is when I log out and log back in, my choice is no longer remembered, I am given the modal pop-up again. 
But I have seen the reason , this is, check the code review to see why this is happening.

File: frontend/src/layouts/MainLayout.tsx [hunk 1]
```diff
@@ -124,17 +125,18 @@ class MainLayout extends React.Component<MainLayoutProps, MainLayoutState> {
                 </Sidebar.Item>
                 <Sidebar.Item
                   active={this.state.activeRoute.includes('/queries/build/')}
-                  collapseId="query-builder"
+                  // collapseId="query-builder"
```
Comment (ThatcherK) [inline frontend/src/layouts/MainLayout.tsx]: Delete this commented out line, line 128

File: frontend/src/pages/Login/Login.tsx [hunk 2]
```diff
@@ -107,13 +107,20 @@ export class Login extends React.Component<LoginProps, LoginState> {
...
         } else if (detail) {
           this.setState({ alert: detail });
         }
       })
       .catch(console.log); // tslint:disable-line
   };
 
+  private savePreference(token: string) {
+    getPreference(token).then((response) => response.json());
+    // .then(({ results }) => {
```
Comment (ThatcherK) [inline frontend/src/pages/Login/Login.tsx]: Remove commented code, not being used

File: frontend/src/pages/QueryBuilderChooser/QueryBuilderChooser.tsx [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+
+type SelectedQueryBuilder = 'basic' | 'advanced';
+
+// eslint-disable-next-line @typescript-eslint/ban-types
+const QueryBuilderChooser: FC<RouteComponentProps> = (props: RouteComponentProps<{}>) => {
+  const [showModal, setShowModal] = useState(true);
+  const [selectedOption, setSelectedOption] = useState<SelectedQueryBuilder>('advanced');
+  const [checked, setChecked] = useState(false);
+  const [choice, setChoice] = useState<string>();
+  const [redirectPage, setRedirectPage] = useState('');
```
Comment (ThatcherK) [inline frontend/src/pages/QueryBuilderChooser/QueryBuilderChooser.tsx]: This variable looks redundant, I don't see it's importance. Why did you add it? I feel like the variable selectedOption is enough, and the redirectPage variable is just a repetition of the same

File: frontend/src/pages/QueryBuilderChooser/QueryBuilderChooser.tsx [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+          method: 'post',
+          url: api.routes.USERPREFERENCE,
+          headers: {
+            'Content-Type': 'application/json',
+            Authorization: `token ${token}`,
+          },
+          data: userPreference,
+        })
+        .then(() => {
+          if (selectedOption === 'basic') {
```
Comment (ThatcherK) [inline frontend/src/pages/QueryBuilderChooser/QueryBuilderChooser.tsx]: This if statement is redundant, not needed

File: frontend/src/pages/QueryBuilderChooser/QueryBuilderChooser.tsx [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+          }
+        })
+        .catch((err) => console.log(err));
+    });
+    if (checked === true) {
+      localForage.setItem(localForageKeys.PREFERENCES, selectedOption);
+    }
+  };
+
+  if (redirectPage || choice) {
```
Comment (ThatcherK) [inline frontend/src/pages/QueryBuilderChooser/QueryBuilderChooser.tsx]: Instead of redirectPage, use selectedOption

File: frontend/src/pages/QueryBuilderChooser/QueryBuilderChooser.tsx [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+    if (redirectPage === 'basic' || choice === 'basic') {
```
Comment (ThatcherK) [inline frontend/src/pages/QueryBuilderChooser/QueryBuilderChooser.tsx]: use selectedOption

File: frontend/src/pages/QueryBuilderChooser/QueryBuilderChooser.tsx [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+        <Modal.Footer>
+          <CheckBox
+            label="Remember choice"
+            checked={checked}
+            onChange={() => {
+              handleChange();
+            }}
+          />
+          <Button
+            onClick={() => {
```
Comment (ThatcherK) [inline frontend/src/pages/QueryBuilderChooser/QueryBuilderChooser.tsx]: Change this to `onClick={handleSave}`

File: frontend/src/pages/QueryBuilderChooser/QueryBuilderChooser.tsx [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+          />
+          <Alert variant="secondary">
+            <p>The Advanced Query Builder</p>
+          </Alert>
+        </Modal.Body>
+        <Modal.Footer>
+          <CheckBox
+            label="Remember choice"
+            checked={checked}
+            onChange={() => {
```
Comment (ThatcherK) [inline frontend/src/pages/QueryBuilderChooser/QueryBuilderChooser.tsx]: Change this to `onChange={handleChange}`

**Review by ThatcherK — CHANGES_REQUESTED** [review]

**Review by edwinmp — CHANGES_REQUESTED** [review]
Change the alert text for the basic option to:
`The original Query Builder. Creates queries using interconnected steps - no SQL knowledge required`

Change the text of the advanced option to:
`An improved UI with a lot more flexibility & options for creating advanced queries. Some of its features require a little SQL knowledge.`

Change the font size of the alert text to `14px`.

Why is the `Go to Query Builder` button purple? It's probably the ONLY purple button in the entire app ... Our colours are red, grey and white. For buttons, red & grey ... choose one. Classes are `btn-danger` or `btn-dark`. Also, Query Builder is 2 words.

Change from `Remember choice` to `Remember my choice`

Change title to `Select Query Builder`

Give the alert the class `p-2` to reduce on the padding so that it isn't too fat.

When a user clicks outside of the modal, instead of simply closing the modal and leaving them on a blank page, select & show the default QB (advanced)

File: core/urls.py [hunk 9]
```diff
@@ -2,6 +2,7 @@
 from knox import views as knox_views
 from core.views import LoginView
 from core import views as core_views
+# from preference.views import PreferenceList, PreferenceDetail
```
Comment (edwinmp) [inline core/urls.py]: Remove

File: core/urls.py [hunk 10]
```diff
@@ -49,6 +50,9 @@
     path('dataset/alias/<int:pk>/', core_views.OperationColumnAlias.as_view()),
     path('dataset/estimate/<int:pk>/', core_views.EstimateQueryTime.as_view()),
     path('dataset/query/', core_views.GetOperationQuery.as_view()),
+    path('preferences/', core_views.PreferenceList.as_view()),
```
Comment (edwinmp) [inline core/urls.py]: Remove - I don't see any use case where we'd fetch all the preferences... remove the view as well if it's doing nothing else. Wasn't part of the requirements in the first place.

File: core/urls.py [hunk 11]
```diff
@@ -49,6 +50,9 @@
...
+    path('preferences/mine/', core_views.UserPreferenceList.as_view()),
+    path('preferences/<int:pk>/', core_views.PreferenceDetail.as_view()),
```
Comment (edwinmp) [inline core/urls.py]: This too ... unless you're using this somewhere, I'd rather you just keep `preferences/mine` as it should be able to handle fetching and updating target user preferences.

File: core/serializers.py [hunk 12]
```diff
@@ -373,16 +374,22 @@ class Meta:
...
+    user = serializers.ReadOnlyField(source='user.username')
+    class Meta:
+        model = Preference
+        fields = ('id', 'preferences', 'global_choice', 'user')
+
 class UserSerializer(serializers.ModelSerializer):
     tag_set = TagSerializer(many=True, read_only=True)
     operation_set = OperationSerializer(many=True, read_only=True)
     review_set = ReviewSerializer(many=True, read_only=True)
+    preferences_field = PreferenceSerializer(many=True, read_only=True)
```
Comment (edwinmp) [inline core/serializers.py]: I'm not sure what this is doing here - you're clearly not using it. Seems like a good idea to fetch and return the user preferences as part of the Login data, but it seems you abandoned that idea half way. It's up to you to pursue it - otherwise, remove this and the `fields` mention on like 392.

File: core/views.py [hunk 13]
```diff
@@ -618,6 +623,34 @@ class UserDetail(generics.RetrieveAPIView):
...
+class UserPreferenceList(generics.ListCreateAPIView):
+    """ Returns a list of preferences for the current and authenticated user """
+    authentication_classes = [TokenAuthentication]
+    permission_classes = (permissions.IsAuthenticatedOrReadOnly & IsOwnerOrReadOnly,)
+    queryset = Preference.objects.all()
+    serializer_class = PreferenceSerializer
+
+    def perform_create(self, serializer):
+        serializer.save(user=self.request.user)
+    def get_queryset(self):
```
Comment (edwinmp) [inline core/views.py:631]: Blank line before this to separate the functions.

File: frontend/src/pages/Login/Login.tsx [hunk 14]
```diff
@@ -107,13 +107,21 @@ export class Login extends React.Component<LoginProps, LoginState> {
...
         }
       })
       .catch(console.log); // tslint:disable-line
   };
 
+  private savePreference(token: string) {
+    getPreference(token)
+      .then((response) => response.json())
+      .then(({ results }) => {
+        localForage.setItem(localForageKeys.PREFERENCES, results[0].preferences);
```
Comment (edwinmp) [inline frontend/src/pages/Login/Login.tsx:121]: You're diving right into `results[0].preferences`. This assumes there will always be results returned. It's safer to check.
```suggestion
        if (results.length) {
            localForage.setItem(localForageKeys.PREFERENCES, results[0].preferences);
        }
```

File: frontend/src/pages/QueryBuilderChooser/QueryBuilderChooser.tsx [hunk 15]
```diff
@@ -0,0 +1,134 @@
+import React, { FC, useState, useEffect } from 'react';
+import * as localForage from 'localforage';
+import { localForageKeys, api } from '../../utils';
+import { Alert, Button, Modal } from 'react-bootstrap';
+import { RouteComponentProps, useHistory } from 'react-router-dom';
+import { ICheckData, IRadio } from '../../components/IRadio';
+import { CheckBox } from '../../components/CheckBox';
+import axios from 'axios';
```
Comment (edwinmp) [inline frontend/src/pages/QueryBuilderChooser/QueryBuilderChooser.tsx]: In your vscode for this file - `CTRL + SHIFT + P` -> then `Organise Imports`

File: frontend/src/pages/QueryBuilderChooser/QueryBuilderChooser.tsx [hunk 16]
```diff
@@ -0,0 +1,134 @@
...
+
+type SelectedQueryBuilder = 'basic' | 'advanced';
+
+// eslint-disable-next-line @typescript-eslint/ban-types
```
Comment (edwinmp) [inline frontend/src/pages/QueryBuilderChooser/QueryBuilderChooser.tsx]: Remove this, unless doing so results in an error ... I don't think so.

File: frontend/src/pages/QueryBuilderChooser/QueryBuilderChooser.tsx [hunk 17]
```diff
@@ -0,0 +1,134 @@
...
+
+  const toggleModal = () => {
+    setShowModal(!showModal);
+    history.push('/queries/build/advanced/');
+  };
+  const onRadioChange = (data: ICheckData) => {
+    setSelectedOption(data.value as SelectedQueryBuilder);
+  };
+
+  const history = useHistory();
```
Comment (edwinmp) [inline frontend/src/pages/QueryBuilderChooser/QueryBuilderChooser.tsx]: This is a hook ... always have your hooks before all other code or you'll run into issues, so move this to where all the other hooks are.

File: frontend/src/pages/QueryBuilderChooser/QueryBuilderChooser.tsx [hunk 18]
```diff
@@ -0,0 +1,134 @@
...
+  useEffect(() => {
+    localForage.getItem<string>(localForageKeys.PREFERENCES).then((key) => {
+      if (key) {
+        setChoice(key || undefined);
+        if (key === 'advanced') {
+          history.push('/queries/build/advanced/');
+        } else {
+          history.push('/queries/build/basic/');
+        }
```
Comment (edwinmp) [inline frontend/src/pages/QueryBuilderChooser/QueryBuilderChooser.tsx]:
```suggestion
        history.push(`/queries/build/${key}/`);
```

File: frontend/src/pages/QueryBuilderChooser/QueryBuilderChooser.tsx [hunk 19]
```diff
@@ -0,0 +1,134 @@
...
+      }
+    });
+  }, []);
+
+  const handleChange = () => {
```
Comment (edwinmp) [inline frontend/src/pages/QueryBuilderChooser/QueryBuilderChooser.tsx]: What's changing? With so many views, be as specific with your naming as possible. As u did on the radio listener, this can be `onCheckboxChange`

File: frontend/src/pages/QueryBuilderChooser/QueryBuilderChooser.tsx [hunk 20]
```diff
@@ -0,0 +1,134 @@
...
+import { CheckBox } from '../../components/CheckBox';
+import axios from 'axios';
+
+type SelectedQueryBuilder = 'basic' | 'advanced';
+
+// eslint-disable-next-line @typescript-eslint/ban-types
+const QueryBuilderChooser: FC<RouteComponentProps> = () => {
+  const [showModal, setShowModal] = useState(true);
+  const [selectedOption, setSelectedOption] = useState<SelectedQueryBuilder>('advanced');
+  const [checked, setChecked] = useState(false);
```
Comment (edwinmp) [inline frontend/src/pages/QueryBuilderChooser/QueryBuilderChooser.tsx:15]: Give more descriptive name here ... `rememberChoice` & `setRememberChoice`

File: frontend/src/pages/QueryBuilderChooser/QueryBuilderChooser.tsx [hunk 21]
```diff
@@ -0,0 +1,134 @@
...
+  const [choice, setChoice] = useState<string>();
```
Comment (edwinmp) [inline frontend/src/pages/QueryBuilderChooser/QueryBuilderChooser.tsx]: You don't need both `selectedOption` & `choice` ... heck, they are practically synonyms ... make do with just one.

File: frontend/src/pages/QueryBuilderChooser/QueryBuilderChooser.tsx [hunk 22]
```diff
@@ -0,0 +1,134 @@
...
+          url: api.routes.USERPREFERENCE,
+          headers: {
+            'Content-Type': 'application/json',
+            Authorization: `token ${token}`,
+          },
+          data: userPreference,
+        })
+        .then(() => {
+          if (selectedOption || choice) {
+            if (selectedOption === 'basic' || choice === 'basic') {
```
Comment (edwinmp) [inline frontend/src/pages/QueryBuilderChooser/QueryBuilderChooser.tsx]: Clear evidence that you only need one of them.

File: frontend/src/pages/QueryBuilderChooser/QueryBuilderChooser.tsx [hunk 23]
```diff
@@ -0,0 +1,134 @@
...
+              history.push('/queries/build/basic/');
+            } else {
+              history.push('/queries/build/advanced/');
+            }
```
Comment (edwinmp) [inline frontend/src/pages/QueryBuilderChooser/QueryBuilderChooser.tsx]: Simplify as I did somewhere up there... single line.

File: frontend/src/pages/QueryBuilderChooser/QueryBuilderChooser.tsx [hunk 24]
```diff
@@ -0,0 +1,134 @@
...
+          }
+        })
+        .catch((err) => console.log(err));
+    });
+    if (checked === true) {
+      localForage.setItem(localForageKeys.PREFERENCES, selectedOption);
```
Comment (edwinmp) [inline frontend/src/pages/QueryBuilderChooser/QueryBuilderChooser.tsx]: Just realising that your endpoint is returning preferences as a single string instead of the JSON it should be... something is not right ... let me go back and check on that model.

Comment (edwinmp) [reply frontend/src/pages/QueryBuilderChooser/QueryBuilderChooser.tsx]: The preferences are NOT supposed to only hold the option for the query builder ... which is why the instructions for this feature where very clear that the preferences should be a JSON field with a default value `{"queryBuilder": "advanced"}` ... that means under the `preferences` field, to get the query builder, you're supposed to parse the JSON and get the value of the `queryBuilder` property ... similary, that's what you update when saving a user preference. You do NOT save the query builder option directly into the preferences field. I hope you're following.

Why is it this way? Cos at some point, I'd like to add more preferences ... perhaps one that changes the UI theme to dark mode. I cannot do that if the preferences field has `basic` or `advanced` as the value. For this example, it'd probably be `{"queryBuilder": "advanced", "theme": "dark"}`.

File: frontend/src/pages/QueryBuilderChooser/QueryBuilderChooser.tsx [hunk 25]
```diff
@@ -0,0 +1,134 @@
...
+        </p>
+      );
+    } else {
+      return (
+        <p style={{ fontSize: 14 }}>
+          An improved UI with a lot more flexibility & options for creating advanced queries. Some
+          of its features require a little SQL knowledge.
+        </p>
+      );
+    }
```
Comment (edwinmp) [inline frontend/src/pages/QueryBuilderChooser/QueryBuilderChooser.tsx]: Rewrite with a single return statement.

File: frontend/src/pages/QueryBuilderChooser/QueryBuilderChooser.tsx [hunk 26]
```diff
@@ -0,0 +1,134 @@
...
+            inline
+            checked={selectedOption === 'advanced'}
+          />
+          <Alert variant="secondary" className="p-2">
+            {alertMessage()}
+          </Alert>
+        </Modal.Body>
+        <Modal.Footer>
+          <CheckBox label="Remember my choice" checked={checked} onChange={handleChange} />
+          <Button onClick={handleSave} variant="btn-danger" className="btn-danger">
```
Comment (edwinmp) [inline frontend/src/pages/QueryBuilderChooser/QueryBuilderChooser.tsx:105]: Ah, I wondered why every time I clicked `Go to Query Builder` a new preference was added to the database ... my user now has like 20 preferences. Clicking this button shouldn't automatically save. It should only save if a user has chosen  `Remember my choice`. Therefore, I suggest you create a different function for this button that automatically navigates to the user's choice without saving IF they haven't chosen to remember their choice. Else, save.

File: frontend/src/utils/api.ts [hunk 27]
```diff
@@ -27,5 +27,7 @@ export const api = {
     DOWNLOAD_SAVED_QUERYSET: '/api/tables/download/{table}/dataset/',
     FETCH_OPERATION: '/api/operations/{id}',
     DATASET_QUERY: '/api/dataset/query/',
+    PREFERENCES: '/api/preferences/',
```
Comment (edwinmp) [inline frontend/src/utils/api.ts]: Remove this ... not used.

File: frontend/src/utils/index.ts [hunk 28]
```diff
@@ -178,3 +179,13 @@ export const getColumnFromName = (
...
   columns.find((col) => col.name === name) as AdvancedQueryColumn;
+
+export const getPreference = (token: string) => {
+  return window.fetch(api.routes.USERPREFERENCE, {
+    method: 'GET',
+    headers: {
+      'Content-Type': 'application/json',
+      Authorization: `token ${token}`,
+    },
+  });
```
Comment (edwinmp) [inline frontend/src/utils/index.ts]: Would be best if you returned the actual JSON here i.e `response.json()` ... abstract as much within this function as you can so that usages of it don't have to do as much ... also, lets see you can do it with `async ... await`.

File: preference/models.py [hunk 29]
```diff
@@ -0,0 +1,9 @@
+from django.db import models
+from core.models import BaseEntity
+
+# Create your models here.
+def default_value():
+    return {"queryBuilder" : "advanced"}
+class Preference(BaseEntity):
```
Comment (edwinmp) [inline preference/models.py:8]: Space before ... separate functions, classes ... any of those kinds of big blocks of code. It makes for cleaner code.

**Review by edwinmp — CHANGES_REQUESTED** [review]
The bug is gone but there is still some work to do.

File: frontend/src/pages/QueryBuilderChooser/QueryBuilderChooser.tsx (hunk 22 above)
Comment (pauline-n) [reply frontend/src/pages/QueryBuilderChooser/QueryBuilderChooser.tsx]: Reason why i have both is so that someone can still be redirected to the querybuilder without the querybuilder of their choice being saved. Then should they select the remember choice checkbox, their choice will be saved and also they will be redirected. However, Let me do away with one so that they have no option but to select the remember choice checkbox.

Comment (edwinmp) [reply frontend/src/pages/QueryBuilderChooser/QueryBuilderChooser.tsx]: Nope... they don't need to be forced to select. You can do this with only one of them without changing the existing behaviour. Also, if a change I've suggested requires you to change the behaviour, DON'T do it till you check with me.

Comment (pauline-n) [reply frontend/src/pages/QueryBuilderChooser/QueryBuilderChooser.tsx]: Noted. I have some that are doing so. I will check in with you in the morning then

File: frontend/src/pages/QueryBuilderChooser/QueryBuilderChooser.tsx (hunk 24 above)
Comment (pauline-n) [reply frontend/src/pages/QueryBuilderChooser/QueryBuilderChooser.tsx]: Am not sure I have understood this well,

Comment (edwinmp) [reply frontend/src/pages/QueryBuilderChooser/QueryBuilderChooser.tsx]: I hope you now understand what I mean.
